File: fastfed-java-sdk/src/main/java/com/sailpoint/fastfed/sdk/providers/FastFedProvider.java
package com.sailpoint.fastfed.sdk.providers;

import com.sailpoint.fastfed.sdk.model.FastFedAuthenticationObject;
import com.sailpoint.fastfed.sdk.model.FastFedProvisioningObject;
import org.json.simple.JSONObject;
import com.sailpoint.fastfed.sdk.model.DuplicateCheckResult;

public interface FastFedProvider {

    String getMetadata();

    void addAuthenticationApplication(FastFedAuthenticationObject fastFedAuthenticationObject) throws Exception;
    void removeAuthenticationApplication(String entityId) throws Exception;
    void addProvisioningApplication(FastFedProvisioningObject fastFedProvisioningObject) throws Exception;
    void removeProvisioningApplication(String entityId) throws Exception;

    DuplicateCheckResult getDuplicateCheckResult(JSONObject providerMetadata);

}

File: fastfed-node-sdk/src/interfaces/fastfed.provider.ts
import {FastFedProvisioningObject} from "../model/fastfedProvisioningObject.model";
import {FastFedAuthenticationObject} from "../model/fastfedAuthenticationObject.model";
import {DuplicateCheckResult} from "../model/duplicateCheckResult.model";

/**
 * Interface that needs to be implemented by a FastFed provider
 */
export interface IFastFedProvider {
    getMetadata(): object;

    addAuthenticationApplication(fastfedAuthenticationObject: FastFedAuthenticationObject): Promise<void>;
    removeAuthenticationApplication(entityId: string): Promise<void>;

    addProvisioningApplication(fastFedProvisioningObject: FastFedProvisioningObject): Promise<void>;
    removeProvisioningApplication(entityId: string): Promise<void>;

    /**
     * Determine if the FastFed metadata will cause a duplication of some functionality  (SSO, provisioning) at the
     * application provider
     */
    getDuplicateCheckResult(idpMetadata: object): Promise<DuplicateCheckResult>;
}

